File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/helpers/TextFileToPointRDD.java
package com.akikanellis.skyspark.api.helpers;

import com.akikanellis.skyspark.api.utils.point.Points;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import java.awt.geom.Point2D;

/**
 * Helper class to convert a text file to a JavaRDD<Point2D>
 */
public class TextFileToPointRDD {

    private final JavaSparkContext sparkContext;

    public TextFileToPointRDD(JavaSparkContext sparkContext) {
        this.sparkContext = sparkContext;
    }

    public JavaRDD<Point2D> getPointRddFromTextFile(String filePath, String delimiter) {
        JavaRDD<String> lines = sparkContext.textFile(filePath);
        return convertToPoints(lines, delimiter);
    }

    private JavaRDD<Point2D> convertToPoints(JavaRDD<String> lines, String delimiter) {
        return lines.map(line -> Points.pointFromTextLine(line, delimiter));
    }
}

File: skyspark-scala/src/main/scala/com/akikanellis/skyspark/scala/api/helpers/TextFileToPointRdd.scala
package com.akikanellis.skyspark.scala.api.helpers

import com.akikanellis.skyspark.scala.api.algorithms.Point
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD

/**
  * Helper for converting a text file to a point RDD. The file should have one point per line.
  */
object TextFileToPointRdd {

  /**
    * Converts the text file to a point RDD.
    *
    * @param sparkContext An active Spark context
    * @param filePath     The full path to the file
    * @param delimiter    Any single character except a period
    * @return The constructed point RDD
    */
  def convert(sparkContext: SparkContext, filePath: String, delimiter: String = " "): RDD[Point] = {
    sparkContext.textFile(filePath)
      .map(pointFromTextLine(_, delimiter))
  }

  private[helpers] def pointFromTextLine(textLine: String, delimiter: String) = {
    val numberArray = textLine
      .trim
      .split(delimiter)
      .map(_.toDouble)
    Point(numberArray: _*)
  }
}
